packages/-ember-data/tests/unit/adapters/rest-adapter/fetch-options-test.js

```diff
@@ -0,0 +1,143 @@
+import { setupTest } from 'ember-qunit';
+import { module, test } from 'qunit';
+import { fetchOptions } from '@ember-data/adapter/rest';
+
+module('unit/adapters/rest-adapter/fetch-options', function(hooks) {
+  setupTest(hooks);
+
+  test("fetchOptions removes undefined query params when method is POST and 'data' is an object", function(assert) {
+    assert.expect(1);
+
+    const dataAsObject = {
+      a: 1,
+      b: undefined,
+      c: 3,
+      d: null,
+      e: 0,
+      f: false,
+    };
+
+    const undefinedQueryStringOptions = {
+      url: 'https://emberjs.com',
+      method: 'POST',
+      data: dataAsObject,
+    };
+
+    let options = fetchOptions(undefinedQueryStringOptions);
+    assert.deepEqual(options.body, '{"a":1,"c":3,"d":null,"e":0,"f":false}');
+  });
+
+  test('fetchOptions sets the request body correctly when the method is not GET or HEAD', function(assert) {
+    assert.expect(3);
+
+    const baseOptions = {
+      url: '/',
+      method: 'POST',
+      data: { a: 1 },
+    };
+
+    // Tests POST method.
+    let options = fetchOptions(baseOptions);
+    assert.equal(options.body, JSON.stringify(baseOptions.data), 'POST request body correctly set');
+
+    // Tests PUT method.
+    baseOptions.method = 'PUT';
+    options = fetchOptions(baseOptions);
+    assert.equal(options.body, JSON.stringify(baseOptions.data), 'PUT request body correctly set');
+
+    // Tests DELETE method.
+    baseOptions.method = 'DELETE';
+    options = fetchOptions(baseOptions);
+    assert.equal(options.body, JSON.stringify(baseOptions.data), 'DELETE request has the correct body');
+  });
+
+  test("fetchOptions sets the request body correctly when the method is POST and 'data' is a string", function(assert) {
+    assert.expect(1);
+
+    // Tests stringified objects.
+    const stringifiedData = JSON.stringify({ a: 1, b: 2 });
+    const optionsWithStringData = {
+      url: 'https://emberjs.com',
+      method: 'POST',
+      data: stringifiedData,
+    };
+
+    let options = fetchOptions(optionsWithStringData);
+    assert.deepEqual(options.body, stringifiedData);
```

Comment:
Hmm, `deepEqual` seems odd here, we are comparing a string to a string, can we use:

```suggestion
    assert.equal(options.body, stringifiedData);
```